src/libstd/sys/windows/stdio.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use prelude::v1::*;
+use io::prelude::*;
+
+use io::{self, Cursor};
+use libc;
+use str;
+use sys::c;
+use sys::handle::Handle;
+use sys::cvt;
+
+struct NoClose(Option<Handle>);
+
+pub struct Stdin {
+    handle: NoClose,
+    utf8: io::Cursor<Vec<u8>>,
+}
+pub struct Stdout(NoClose);
+pub struct Stderr(NoClose);
+
+fn get(handle: libc::DWORD) -> io::Result<NoClose> {
+    let handle = unsafe { c::GetStdHandle(handle) };
+    if handle == libc::INVALID_HANDLE_VALUE {
+        Err(io::Error::last_os_error())
+    } else if handle.is_null() {
+        Err(io::Error::new(io::ErrorKind::Other,
+                           "no stdio handle available for this process", None))
+    } else {
+        Ok(NoClose::new(handle))
+    }
+}
+
+fn write(handle: libc::HANDLE, data: &[u8]) -> io::Result<usize> {
+    let utf16 = match str::from_utf8(buf).ok() {
+        Some(utf8) => utf8.utf16_units().collect::<Vec<u16>>(),
+        None => return Err(invalid_encoding()),
+    };
+    let mut written: DWORD = 0;
+    try!(cvt(unsafe {
+        WriteConsoleW(handle,
+                      utf16.as_ptr() as LPCVOID,
+                      utf16.len() as u32,
+                      &mut written as LPDWORD,
+                      ptr::null_mut())
+    }));
+
+    // FIXME if this only partially writes the utf16 buffer then we need to
+    //       figure out how many bytes of `data` were actually written
+    assert_eq!(written as usize, utf16.len());
```

Comment:
Unfortunately the relationship of `buf.len()` and `utf16.len()` aren't that correlated I think. If `buf` is all ASCII then I think the lengths are the same, but due to variable-width-encodings of characters I think the precise lengths don't have that much correlation.

Put another way, if we wrote N/M u16 elements of `utf16` then we would have to re-encode `&utf16[..N]` as UTF-8 to figure out how many bytes were written.

I... would love a better way to do this though! I'm pretty uncomfortable with how it's currently implemented.